src/umpire/strategy/ResourceAwarePool.cpp

```diff
@@ -0,0 +1,549 @@
+// Copyright (c) 2016-24, Lawrence Livermore National Security, LLC and Umpire
+// project contributors. See the COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (MIT)
+//////////////////////////////////////////////////////////////////////////////
+#include "umpire/strategy/ResourceAwarePool.hpp"
+
+#include "umpire/Allocator.hpp"
+#include "umpire/event/event.hpp"
+#include "umpire/strategy/PoolCoalesceHeuristic.hpp"
+#include "umpire/strategy/mixins/AlignedAllocation.hpp"
+#include "umpire/util/Macros.hpp"
+#include "umpire/util/memory_sanitizers.hpp"
+
+namespace umpire {
+namespace strategy {
+
+ResourceAwarePool::ResourceAwarePool(const std::string& name, int id, Allocator allocator,
+                                     const std::size_t first_minimum_pool_allocation_size,
+                                     const std::size_t next_minimum_pool_allocation_size, std::size_t alignment,
+                                     PoolCoalesceHeuristic<ResourceAwarePool> should_coalesce) noexcept
+    : AllocationStrategy{name, id, allocator.getAllocationStrategy(), "ResourceAwarePool"},
+      mixins::AlignedAllocation{alignment, allocator.getAllocationStrategy()},
+      m_should_coalesce{should_coalesce},
+      m_first_minimum_pool_allocation_size{first_minimum_pool_allocation_size},
+      m_next_minimum_pool_allocation_size{next_minimum_pool_allocation_size}
+{
+  UMPIRE_LOG(Debug, " ( "
+                        << "name=\"" << name << "\""
+                        << ", id=" << id << ", allocator=\"" << allocator.getName() << "\""
+                        << ", first_minimum_pool_allocation_size=" << m_first_minimum_pool_allocation_size
+                        << ", next_minimum_pool_allocation_size=" << m_next_minimum_pool_allocation_size
+                        << ", alignment=" << alignment << " )");
+}
+
+ResourceAwarePool::~ResourceAwarePool()
+{
+  UMPIRE_LOG(Debug, "Releasing free blocks to device");
+  m_is_destructing = true;
+  release();
+}
+
+void* ResourceAwarePool::allocate(std::size_t UMPIRE_UNUSED_ARG(bytes))
+{
+  void* ptr{nullptr};
+  UMPIRE_ERROR(
+      runtime_error,
+      fmt::format("The ResourceAwarePool requires a Camp resource. See "
+                  "https://umpire.readthedocs.io/en/develop/sphinx/cookbook/resource_aware_pool.html for more info."));
+  return ptr;
+}
+
+void* ResourceAwarePool::allocate_resource(camp::resources::Resource r, std::size_t bytes)
+{
+  UMPIRE_LOG(Debug, "(bytes=" << bytes << ")");
+  const std::size_t rounded_bytes{aligned_round_up(bytes)};
+  const auto& best = m_free_map.lower_bound(rounded_bytes);
```

Comment:
This could be moved down to line 85 (no need to search the free map if there is a usable chunk in the pending map).

Reply:
Done!